File: src/test/java/application/SimulatorApplicationTest.java
package application;

import org.assertj.core.api.Assertions;
import org.junit.Test;
import util.MockClock;
import util.MockPrinter;

public class SimulatorApplicationTest {

    @Test
    public void ShouldValidSimulationOutput() {
        // Given
        MockPrinter printer = new MockPrinter();
        SimulatorApplication app = new SimulatorApplication(printer, new MockClock("17:53"));

        // When
        app.run();

        // Then
        Assertions.assertThat(printer.getOut().toString()).isEqualTo("\n" +
                "\n" +
                "╔══════════════════════════════════════════════════════════╗\n" +
                "║                           City                           ║\n" +
                "╚══════════════════════════════════════════════════════════╝\n" +
                "                                                 Sillingy ✓ \n" +
                "                                                   Annecy ✓ \n" +
                "                                                   Epagny ✓ \n" +
                "                                               Metz-Tessy ✓ \n" +
                "                                                   Seynod ✓ \n" +
                "\n" +
                "\n" +
                "╔══════════════════════════════════════════════════════════╗\n" +
                "║                           Road                           ║\n" +
                "╚══════════════════════════════════════════════════════════╝\n" +
                "                               Sillingy ══[1min]=> Epagny ✓ \n" +
                "                             Epagny ══[3min]=> Metz-Tessy ✓ \n" +
                "                            Metz-Tessy ══[12min]=> Annecy ✓ \n" +
                "                             Metz-Tessy ══[4min]=> Seynod ✓ \n" +
                "                              Sillingy ══[11min]=> Seynod ✓ \n" +
                "                              Sillingy ══[19min]=> Annecy ✓ \n" +
                "                                 Annecy ══[9min]=> Seynod ✓ \n" +
                "\n" +
                "\n" +
                "╔══════════════════════════════════════════════════════════╗\n" +
                "║                           Map                            ║\n" +
                "╚══════════════════════════════════════════════════════════╝\n" +
                "┌──────────────────────────────────────────────────────────┐\n" +
                "│          Algorithm: Home made Djikstra by Tibo           │\n" +
                "├──────────────────┬───────────────────────────────────────┤\n" +
                "│      Cities      │                 Roads                 │\n" +
                "├──────────────────┼───────────────────────────────────────┤\n" +
                "│Sillingy          │Sillingy ══[1min]=> Epagny             │\n" +
                "│Annecy            │Epagny ══[3min]=> Metz-Tessy           │\n" +
                "│Epagny            │Metz-Tessy ══[12min]=> Annecy          │\n" +
                "│Metz-Tessy        │Metz-Tessy ══[4min]=> Seynod           │\n" +
                "│Seynod            │Sillingy ══[11min]=> Seynod            │\n" +
                "│                  │Sillingy ══[19min]=> Annecy            │\n" +
                "│                  │Annecy ══[9min]=> Seynod               │\n" +
                "└──────────────────┴───────────────────────────────────────┘\n" +
                "\n" +
                "\n" +
                "╔══════════════════════════════════════════════════════════╗\n" +
                "║                          Route                           ║\n" +
                "╚══════════════════════════════════════════════════════════╝\n" +
                "╔══════════════════════════════════════════════════════════╗\n" +
                "║                   Sillingy ==> Annecy                    ║\n" +
                "╠═════════════════════════════╦════════════════════════════╣\n" +
                "║       Arrived: 17:53        ║      Duration: 16 min      ║\n" +
                "╠═════════════════════════════╩════════════════════════════╣\n" +
                "║1. Sillingy ══[1min]=> Epagny                             ║\n" +
                "║2. Epagny ══[3min]=> Metz-Tessy                           ║\n" +
                "║3. Metz-Tessy ══[12min]=> Annecy                          ║\n" +
                "╚══════════════════════════════════════════════════════════╝\n");
    }
}

File: src/test/java/application/SimulatorApplicationTest.kt
package application

import application.print.Clockable
import application.print.Printable
import org.assertj.core.api.Assertions
import org.junit.Test

class SimulatorApplicationTest {

    @Test
    fun shouldDisplayTheCorrectOutput() {
        // Given
        val printer = TestPrinter()
        val simulator = SimulatorApplication(printer, TestClock())

        // When
        simulator.run()

        // Then
        Assertions.assertThat(printer.out.toString()).isEqualTo("""


╔══════════════════════════════════════════════════════════╗
║                           City                           ║
╚══════════════════════════════════════════════════════════╝
                                                 Sillingy ✓ 
                                                   Annecy ✓ 
                                                   Epagny ✓ 
                                               Metz-Tessy ✓ 
                                                   Seynod ✓ 


╔══════════════════════════════════════════════════════════╗
║                           Road                           ║
╚══════════════════════════════════════════════════════════╝
                               Sillingy ══[1min]=> Epagny ✓ 
                             Epagny ══[3min]=> Metz-Tessy ✓ 
                            Metz-Tessy ══[12min]=> Annecy ✓ 
                             Metz-Tessy ══[4min]=> Seynod ✓ 
                              Sillingy ══[11min]=> Seynod ✓ 
                              Sillingy ══[19min]=> Annecy ✓ 
                                 Annecy ══[9min]=> Seynod ✓ 


╔══════════════════════════════════════════════════════════╗
║                           Map                            ║
╚══════════════════════════════════════════════════════════╝
┌──────────────────┬───────────────────────────────────────┐
│      Cities      │                 Roads                 │
├──────────────────┼───────────────────────────────────────┤
│Sillingy          │Sillingy ══[1min]=> Epagny             │
│Annecy            │Epagny ══[3min]=> Metz-Tessy           │
│Epagny            │Metz-Tessy ══[12min]=> Annecy          │
│Metz-Tessy        │Metz-Tessy ══[4min]=> Seynod           │
│Seynod            │Sillingy ══[11min]=> Seynod            │
│                  │Sillingy ══[19min]=> Annecy            │
│                  │Annecy ══[9min]=> Seynod               │
└──────────────────┴───────────────────────────────────────┘


╔══════════════════════════════════════════════════════════╗
║                          Route                           ║
╚══════════════════════════════════════════════════════════╝
╔══════════════════════════════════════════════════════════╗
║                   Sillingy ==> Annecy                    ║
╠═════════════════════════════╦════════════════════════════╣
║       Arrived: 17:17        ║      Duration: 16 min      ║
╠═════════════════════════════╩════════════════════════════╣
║1. Sillingy ══[1min]=> Epagny                             ║
║2. Epagny ══[3min]=> Metz-Tessy                           ║
║3. Metz-Tessy ══[12min]=> Annecy                          ║
╚══════════════════════════════════════════════════════════╝

        """.trimIndent())
    }

    class TestPrinter : Printable {
        val out = StringBuilder()

        override fun println(text: String) {
            out.appendln(text)
        }
    }

    class TestClock : Clockable {
        override fun nowInHHmm(addMinutes: Int) = "17:17"

    }
}